File: leetcode/top_interview_150/125. Valid Palindrome/main.go
// You can edit this code!
// Click here and start typing.
package main

import (
	"fmt"
	"regexp"
	"strings"
)			

func removeNonAlphanumericChars(s string) string {
    reg, _ := regexp.Compile("[^a-zA-Z0-9]+")
	processedString := reg.ReplaceAllString(s, "")
    return processedString
}

func isPalindrome(s string) bool {
    lower := strings.ToLower(s)
	trimmed := strings.TrimSpace(lower)
	nonAlphCh := removeNonAlphanumericChars(trimmed)

	left := 0
	right := len(nonAlphCh) - 1

	for left < len(nonAlphCh) || right > 0 {
		if nonAlphCh[left] != nonAlphCh[right] {
			return  false
		}
		left++; right--
	}

	return true
}

func main() {
	fmt.Println(isPalindrome("A man, a plan, a canal: Panama"))
	fmt.Println(isPalindrome("race a car"))
	fmt.Println(isPalindrome(" "))
}

File: leetcode/top_interview_150/125. Valid Palindrome/main.py
import re

remove_non_alch_ch = lambda s: re.sub(r'[^a-zA-Z0-9]', '', s) 

class Solution:
    def isPalindrome(self, s: str) -> bool:
        s = s.lower()
        s = s.strip()
        s = remove_non_alch_ch(s)
        l = 0
        r = len(s) - 1
        while l < len(s) or r > 0:
            if s[l] != s[r]:
                return False
            l+=1
            r-=1
        return True



if __name__ == "__main__":
    print(Solution().isPalindrome("A man, a plan, a canal: Panama"))
    print(Solution().isPalindrome("race a car"))
    print(Solution().isPalindrome(" "))